File: apps/solver/feasibility.py
"""Independent feasibility check of one dispatch scenario (review F04).

Every scenario the engine returns - the route search's own plan, a post-solve candidate, or a
fallback kept when the post-solve stage did not run - is re-checked here against the REQUEST and
the road matrix, from the emitted minutes alone. Nothing here trusts the engine's own timing code
(load_repack.timing_ok, _build_scenario): the rules are re-derived from the request, so a timing
bug or a fallback path shows up as a violation instead of a timetable no truck can drive.

Rules (all times in whole minutes, as the scenario reports them; TOL_MIN absorbs the rounding of
seconds to minutes in dispatch_solver._min_of):

* capacity: a load's cases (and kg, when the truck has a payload) from the request's stops;
* hard receiving windows: service starts inside the window;
* travel: each service start is at least the previous departure + the drive time (matrix), and
  the truck is back no earlier than the last departure + the drive back;
* unloading: departure - service start == the service time that was sent;
* turnaround: a load departs no earlier than the previous load's return + reload + loading time
  per case x ITS cases - also after the truck's last frozen (locked / dispatched) load;
* the truck day: first departure after shift start, depot opening and the truck's availability;
  every return before the depot closes and the truck's availability ends; first departure (the
  first frozen one when there is one) -> last return within the shift maximum;
* loads per truck: frozen + new <= max trips; new loads never overlap a frozen one and carry the
  load numbers after it.

Violation messages are for the dispatcher: plain words, clock times, minutes short.
"""
from __future__ import annotations

import logging
from typing import TYPE_CHECKING

from dispatch_models import (
    DAY_MIN,
    DispatchRequest,
    DispatchScenario,
    DispatchStop,
    FeasibilityReport,
    FeasibilityViolation,
)

if TYPE_CHECKING:  # pragma: no cover
    from providers import MatrixResult

log = logging.getLogger("routeiq.dispatch.feasibility")

TOL_MIN = 1  # minutes: every emitted time is rounded to a whole minute
KG_TOL = 0.05
CHECK_VERSION = 1


def _hhmm(m: float | int | None) -> str:
    if m is None:
        return "--:--"
    m = int(round(m))
    return f"{(m // 60) % 24:02d}:{m % 60:02d}" + ("+1" if m >= DAY_MIN else "")


def _short(x: float) -> float:
    return round(x, 1)


def check_scenario(
    req: DispatchRequest,
    sc: DispatchScenario,
    *,
    solvable: list[DispatchStop] | None = None,
    mx: "MatrixResult | None" = None,
    timing: str = "EXACT",
) -> FeasibilityReport:
    """Check ``sc`` against ``req``. ``solvable`` + ``mx``: the matrix the scenario was planned on
    (node 0 = depot, node k + 1 = solvable[k]); without them drive times are not re-checked."""
    cfg = req.config
    trucks = {t.id: t for t in req.trucks}
    stops = {s.stop_id: s for s in req.stops}
    node_of = {s.stop_id: k + 1 for k, s in enumerate(solvable)} if (solvable is not None and mx is not None) else None
    out: list[FeasibilityViolation] = []

    def add(code: str, message: str, *, truck_id=None, load_no=None, stop_id=None, short=None) -> None:
        out.append(FeasibilityViolation(code=code, message=message, truck_id=truck_id, load_no=load_no, stop_id=stop_id,  # type: ignore[arg-type]
                                        short_by_min=_short(short) if short is not None else None))

    def drive_min(a: str | None, b: str | None) -> float | None:
        """Drive time in minutes between two stops (None = depot); None when not checkable."""
        if node_of is None or mx is None:
            return None
        i = 0 if a is None else node_of.get(a)
        j = 0 if b is None else node_of.get(b)
        if i is None or j is None:
            return None
        return mx.duration_s[i][j] / 60.0

    def turnaround(cases: int) -> float:
        return cfg.reload_min + cfg.loading_min_per_case * cases

    by_truck: dict[str, list] = {}
    for ld in sc.loads:
        by_truck.setdefault(ld.truck_id, []).append(ld)

    depot_close = req.depot.close_min if req.depot.close_min > 0 else DAY_MIN
    for tid, loads in by_truck.items():
        t = trucks.get(tid)
        if t is None:
            add("UNKNOWN_TRUCK", f"Truck {tid} is not in the request.", truck_id=tid)
            continue
        code = t.code or t.id
        loads = sorted(loads, key=lambda l: (l.load_no, l.depart_min))
        frozen = sorted(t.frozen_trips, key=lambda f: f.load_no)
        max_trips = t.max_trips or cfg.max_trips_per_truck

        # Loads per day and load numbers (new loads follow the frozen ones).
        if len(frozen) + len(loads) > max_trips:
            add("TRIPS", f"Truck {code} has {len(frozen) + len(loads)} loads ({len(frozen)} locked or out); at most {max_trips} per day.",
                truck_id=tid)
        expected = list(range(len(frozen) + 1, len(frozen) + len(loads) + 1))
        if [l.load_no for l in loads] != expected:
            add("LOAD_NUMBER", f"Truck {code} load numbers {[l.load_no for l in loads]} do not follow its "
                               f"{len(frozen)} locked or dispatched load(s) (expected {expected}).", truck_id=tid)

        # The truck day's bounds.
        earliest = max(cfg.shift_start_min, req.depot.open_min, t.available_from_min or 0)
        why_earliest = ("the shift start" if earliest == cfg.shift_start_min else
                        "the depot opening" if earliest == req.depot.open_min else "the truck's availability")
        avail_to = t.available_to_min if t.available_to_min is not None else DAY_MIN * 2
        anchor = min([f.depart_min for f in frozen] + [l.depart_min for l in loads])
        last_return = max(l.return_min for l in loads)
        if last_return - anchor > cfg.shift_max_min + TOL_MIN:
            add("SHIFT_LIMIT", f"Truck {code} is out from {_hhmm(anchor)} to {_hhmm(last_return)}: longer than the "
                               f"{cfg.shift_max_min // 60}h{cfg.shift_max_min % 60:02d} shift maximum.",
                truck_id=tid, short=last_return - anchor - cfg.shift_max_min)

        prev_return: int | None = max((f.return_min for f in frozen), default=None)
        prev_what = "its last locked or dispatched load"
        for ld in loads:
            lno = ld.load_no
            # Capacity: from the request's stops, not from the load's own totals.
            known = [stops.get(st.stop_id) for st in ld.stops]
            for st, s in zip(ld.stops, known):
                if s is None:
                    add("UNKNOWN_STOP", f"Stop {st.stop_id} on {code} load {lno} is not in the request.", truck_id=tid,
                        load_no=lno, stop_id=st.stop_id)
            cases = sum(s.demand_cases for s in known if s is not None)
            kg = sum(s.demand_kg for s in known if s is not None)
            if ld.cases != cases:
                add("LOAD_TOTALS", f"{code} load {lno} records {ld.cases} cases but its stops add up to {cases}.",
                    truck_id=tid, load_no=lno)
            if cases > t.capacity_cases:
                add("CAPACITY_CASES", f"{code} load {lno} carries {cases} cases; the truck takes {t.capacity_cases}.",
                    truck_id=tid, load_no=lno, short=cases - t.capacity_cases)
            if t.capacity_kg > 0 and kg > t.capacity_kg + KG_TOL:
                add("CAPACITY_KG", f"{code} load {lno} weighs {kg:.0f} kg; the truck's payload is {t.capacity_kg:.0f} kg.",
                    truck_id=tid, load_no=lno, short=kg - t.capacity_kg)

            # Departure: after the truck is ready (turnaround) and inside its day.
            if prev_return is not None:
                ready = prev_return + turnaround(ld.cases)
                if ld.depart_min < ready - TOL_MIN:
                    add("TURNAROUND",
                        f"{code} load {lno} leaves at {_hhmm(ld.depart_min)}, but after {prev_what} (back {_hhmm(prev_return)}) "
                        f"the truck needs {turnaround(ld.cases):g} min to reload and load {ld.cases} cases: ready {_hhmm(ready)}.",
                        truck_id=tid, load_no=lno, short=ready - ld.depart_min)
            if ld.depart_min < earliest - TOL_MIN:
                add("EARLY_DEPARTURE", f"{code} load {lno} leaves at {_hhmm(ld.depart_min)}, before {why_earliest} ({_hhmm(earliest)}).",
                    truck_id=tid, load_no=lno, short=earliest - ld.depart_min)
            for f in frozen:
                if ld.depart_min < f.return_min and f.depart_min < ld.return_min:
                    add("FROZEN_OVERLAP", f"{code} load {lno} ({_hhmm(ld.depart_min)}-{_hhmm(ld.return_min)}) overlaps its locked or "
                                          f"dispatched load {f.load_no} ({_hhmm(f.depart_min)}-{_hhmm(f.return_min)}).",
                        truck_id=tid, load_no=lno)

            # The stops: drive, window, unloading.
            prev_stop: str | None = None
            prev_dep = ld.depart_min
            for st in ld.stops:
                s = stops.get(st.stop_id)
                leg = drive_min(prev_stop, st.stop_id)
                if leg is not None and st.service_start_min < prev_dep + leg - TOL_MIN:
                    add("TRAVEL", f"{code} load {lno}: {st.stop_id} starts at {_hhmm(st.service_start_min)}, but the drive from "
                                  f"{'the depot' if prev_stop is None else prev_stop} ({_hhmm(prev_dep)}) takes {leg:.0f} min.",
                        truck_id=tid, load_no=lno, stop_id=st.stop_id, short=prev_dep + leg - st.service_start_min)
                if s is not None:
                    hs = s.hard_start_min or 0
                    he = s.hard_end_min if s.hard_end_min is not None else DAY_MIN * 2
                    if not (hs <= st.service_start_min <= he) or not st.hard_window_ok:
                        add("HARD_WINDOW", f"{code} load {lno}: {st.stop_id} is served at {_hhmm(st.service_start_min)}, outside its "
                                           f"receiving hours {_hhmm(s.hard_start_min)}-{_hhmm(s.hard_end_min)}.",
                            truck_id=tid, load_no=lno, stop_id=st.stop_id,
                            short=max(hs - st.service_start_min, st.service_start_min - he, 0))
                    if st.departure_min - st.service_start_min != s.service_min:
                        add("SERVICE_TIME", f"{code} load {lno}: {st.stop_id} is given {st.departure_min - st.service_start_min} min to "
                                            f"unload; its service time is {s.service_min} min.",
                            truck_id=tid, load_no=lno, stop_id=st.stop_id,
                            short=s.service_min - (st.departure_min - st.service_start_min))
                prev_stop, prev_dep = st.stop_id, st.departure_min
            back = drive_min(prev_stop, None) if ld.stops else None
            if back is not None and ld.return_min < prev_dep + back - TOL_MIN:
                add("RETURN", f"{code} load {lno} is back at {_hhmm(ld.return_min)}, but the drive back from the last stop "
                              f"({_hhmm(prev_dep)}) takes {back:.0f} min.",
                    truck_id=tid, load_no=lno, short=prev_dep + back - ld.return_min)
            if ld.return_min > depot_close + TOL_MIN:
                add("DEPOT_CLOSE", f"{code} load {lno} is back at {_hhmm(ld.return_min)}, after the depot closes ({_hhmm(depot_close)}).",
                    truck_id=tid, load_no=lno, short=ld.return_min - depot_close)
            if ld.return_min > avail_to + TOL_MIN:
                add("TRUCK_AVAILABILITY", f"{code} load {lno} is back at {_hhmm(ld.return_min)}, after the truck's availability ends "
                                          f"({_hhmm(avail_to)}).", truck_id=tid, load_no=lno, short=ld.return_min - avail_to)
            prev_return, prev_what = ld.return_min, f"load {lno}"

    status = "VIOLATED" if out else "VERIFIED"
    return FeasibilityReport(status=status, timing="EXACT" if timing == "EXACT" else "ESTIMATED", violations=out,  # type: ignore[arg-type]
                             checked_at_version=CHECK_VERSION, travel_checked=node_of is not None)


def safe_check(req: DispatchRequest, sc: DispatchScenario, **kw) -> FeasibilityReport:
    """check_scenario that never fails the request: an error in the check itself gives an
    UNVERIFIED report (not dispatchable on the web) and a log line."""
    try:
        return check_scenario(req, sc, **kw)
    except Exception as exc:  # noqa: BLE001
        log.exception("feasibility check failed for %s: %s", sc.name, exc)
        return FeasibilityReport(status="UNVERIFIED", timing="EXACT" if kw.get("timing", "EXACT") == "EXACT" else "ESTIMATED",
                                 checked_at_version=CHECK_VERSION, travel_checked=False,
                                 note="The timetable check could not run for this plan.")

File: apps/web/lib/dispatch/feasibility.ts
/**
 * The dispatch feasibility gate (review F04, part B). Pure: no database access here - the plan
 * service loads the rows (loadFeasibilityInput in plan-service.ts) and stores the result in
 * RunPlan.feasibilityJson.
 *
 * Checked per TRUCK-DAY over every load of the plan version, frozen copies included, from the
 * plan's own facts - the load and stop snapshots and the optimizer inputs - never from today's
 * master data:
 * - cases per load (from its stops) vs the truck's capacity;
 * - physical kg per load (a split portion's own kg, else the order's) vs the payload; lines
 *   with no weight (0 kg) are reported as KG_UNKNOWN (a warning: the dispatcher accepted them
 *   explicitly at optimize, review F02);
 * - service starts inside the hard receiving window, and the stored "within hours" flag;
 * - turnaround: each load leaves after the previous load's return + reload + loading time per
 *   case x its cases (the rules the later load was planned with);
 * - the truck day: first departure after shift start / depot opening / truck availability,
 *   returns before the depot closes / availability ends, first departure -> last return within
 *   the shift maximum, loads per day;
 * - and the solver's own report for the option in use (apps/solver/feasibility.py): VIOLATED
 *   blocks its trucks; a missing report (an option from before the report existed, or a solver
 *   older than the web) is UNKNOWN - such a plan is blocked only on a concrete violation found
 *   here ("LEGACY_STRUCTURAL_ONLY").
 * Times are whole minutes with a 1-minute tolerance (the solver rounds seconds to minutes).
 *
 * A problem on a load that has already left (DISPATCHED / COMPLETED) is history: it is reported
 * as a warning, never blocks, so it can never lock the truck's later loads for good. A turnaround
 * problem belongs to the LATER load (the one that leaves too early), which a re-plan can move.
 *
 * LOCK, LOADING and DISPATCH of a load are refused while its truck-day is not OK
 * (plan-service.changeStatusTx), unless the operator switch FEASIBILITY_GATE=warn is set.
 */
import { createHash } from 'node:crypto';
import type { FeasibilityReport } from '@routeiq/shared-types';
import type { PlanRules } from './snapshots';

export const FEASIBILITY_VERSION = 1;
export const TOL_MIN = 1;
const KG_TOL = 0.05;

export type WebViolationCode =
  | 'CAPACITY_CASES'
  | 'CAPACITY_KG'
  | 'KG_UNKNOWN'
  | 'HARD_WINDOW'
  | 'STOP_TIMES'
  | 'TURNAROUND'
  | 'EARLY_DEPARTURE'
  | 'DEPOT_CLOSE'
  | 'TRUCK_AVAILABILITY'
  | 'SHIFT_LIMIT'
  | 'TRIPS'
  | 'SOLVER';

export interface PlanViolation {
  code: WebViolationCode | string;
  /** BLOCK: LOCK / LOADING / DISPATCH refused for the truck. WARN: shown, not blocking. */
  severity: 'BLOCK' | 'WARN';
  source: 'WEB' | 'SOLVER';
  truckId: string | null;
  truckCode: string | null;
  loadId: string | null;
  loadNo: number | null;
  message: string;
  shortBy?: number | null;
}

export type TruckTiming = 'VERIFIED' | 'VIOLATED' | 'UNVERIFIED' | 'STRUCTURAL_ONLY';

export interface PlanFeasibility {
  v: number;
  /** No truck-day is blocked. */
  ok: boolean;
  status: TruckTiming;
  /** SOLVER_AND_WEB: the solver checked the option in use too. LEGACY_STRUCTURAL_ONLY: no solver
   * report (an older option or solver) - blocked only on a concrete violation found here. */
  source: 'SOLVER_AND_WEB' | 'LEGACY_STRUCTURAL_ONLY';
  solverStatus: 'VERIFIED' | 'VIOLATED' | 'UNVERIFIED' | 'UNKNOWN';
  solverTiming: 'EXACT' | 'ESTIMATED' | null;
  trucks: Record<string, { truckCode: string; status: TruckTiming; ok: boolean; blocking: number; warnings: number }>;
  violations: PlanViolation[];
  inputHash: string;
  checkedAt: string;
}

export interface FeasStop {
  orderId: string;
  sequence: number;
  label: string;
  cases: number;
  kg: number;
  /** Some of these cases have no weight (0 kg). */
  kgUnknown: boolean;
  etaMin: number | null;
  serviceStartMin: number | null;
  departureMin: number | null;
  hardWindowOk: boolean | null;
  /** The receiving window the stop was planned with (snapshot); undefined = not known. */
  hardStartMin?: number | null;
  hardEndMin?: number | null;
}

export interface FeasLoad {
  id: string;
  truckId: string;
  truckCode: string;
  loadNo: number;
  /** DISPATCHED / COMPLETED: its problems are history (warnings, never blocking). */
  onRoad: boolean;
  departMin: number;
  returnMin: number;
  cases: number;
  weightKg: number;
  /** The truck the load was planned on (snapshot, else the optimizer request); null = not known. */
  capacity: { cases: number; kg: number } | null;
  /** The rules the load was timed with; null = not known (no turnaround / shift check). */
  rules: PlanRules | null;
  stops: FeasStop[];
}

export interface FeasibilityInput {
  scenarioId: string | null;
  /** The solver's report for the option in use; undefined / null = none sent (UNKNOWN). */
  solver: FeasibilityReport | null | undefined;
  loads: FeasLoad[];
}

const hhmm = (m: number) => {
  const r = Math.round(m);
  return `${String(Math.floor(r / 60) % 24).padStart(2, '0')}:${String(((r % 60) + 60) % 60).padStart(2, '0')}${r >= 1440 ? '+1' : ''}`;
};
const r1 = (x: number) => Math.round(x * 10) / 10;

export function inputHash(input: FeasibilityInput): string {
  const loads = [...input.loads]
    .sort((a, b) => a.id.localeCompare(b.id))
    .map((l) => [
      l.id, l.truckId, l.loadNo, l.onRoad, l.departMin, l.returnMin, l.cases, r1(l.weightKg), l.capacity?.cases ?? null, l.capacity?.kg ?? null,
      l.rules ? [l.rules.shiftStartMin, l.rules.shiftMaxMin, l.rules.reloadMin, l.rules.loadingMinPerCase, l.rules.maxTrips, l.rules.depotOpenMin, l.rules.depotCloseMin, l.rules.availableFromMin, l.rules.availableToMin] : null,
      [...l.stops]
        .sort((a, b) => a.sequence - b.sequence || a.orderId.localeCompare(b.orderId))
        .map((s) => [s.orderId, s.sequence, s.cases, r1(s.kg), s.kgUnknown, s.etaMin, s.serviceStartMin, s.departureMin, s.hardWindowOk, s.hardStartMin ?? null, s.hardEndMin ?? null]),
    ]);
  const solver = input.solver ? [input.solver.status, input.solver.timing, input.solver.violations.map((v) => [v.code, v.truck_id ?? null, v.load_no ?? null])] : null;
  return createHash('sha256').update(JSON.stringify({ v: FEASIBILITY_VERSION, s: input.scenarioId, solver, loads })).digest('hex');
}

/** The timetable check of one plan version (see the top of this file). */
export function checkPlanFeasibility(input: FeasibilityInput, now: Date = new Date()): PlanFeasibility {
  const out: PlanViolation[] = [];
  const byTruck = new Map<string, FeasLoad[]>();
  for (const l of input.loads) byTruck.set(l.truckId, [...(byTruck.get(l.truckId) ?? []), l]);
  const codeOf = new Map(input.loads.map((l) => [l.truckId, l.truckCode]));
  const onRoad = new Set(input.loads.filter((l) => l.onRoad).map((l) => l.id));
  const history = ' (the load has already left: shown for the record, it does not block)';
  const v = (x: Omit<PlanViolation, 'source' | 'severity'> & { severity?: PlanViolation['severity'] }) =>
    out.push(
      x.loadId && onRoad.has(x.loadId) && (x.severity ?? 'BLOCK') === 'BLOCK'
        ? { source: 'WEB', ...x, severity: 'WARN', message: x.message + history }
        : { severity: 'BLOCK', source: 'WEB', ...x },
    );

  for (const [truckId, list] of byTruck) {
    const loads = [...list].sort((a, b) => a.loadNo - b.loadNo || a.departMin - b.departMin);
    const code = loads[0].truckCode;
    const at = (l: FeasLoad) => ({ truckId, truckCode: code, loadId: l.id, loadNo: l.loadNo });
    for (const l of loads) {
      const cases = l.stops.reduce((a, s) => a + s.cases, 0);
      const kg = r1(l.stops.reduce((a, s) => a + s.kg, 0));
      if (l.capacity) {
        if (cases > l.capacity.cases) {
          v({ ...at(l), code: 'CAPACITY_CASES', message: `${code} load ${l.loadNo} carries ${cases} cases; the truck takes ${l.capacity.cases}.`, shortBy: cases - l.capacity.cases });
        }
        if (l.capacity.kg > 0 && kg > l.capacity.kg + KG_TOL) {
          v({ ...at(l), code: 'CAPACITY_KG', message: `${code} load ${l.loadNo} weighs ${Math.round(kg)} kg; the truck's payload is ${Math.round(l.capacity.kg)} kg.`, shortBy: r1(kg - l.capacity.kg) });
        }
        const unknown = l.stops.filter((s) => s.kgUnknown);
        if (l.capacity.kg > 0 && unknown.length) {
          v({
            ...at(l),
            severity: 'WARN',
            code: 'KG_UNKNOWN',
            message: `${code} load ${l.loadNo}: ${unknown.length} stop(s) have cases with no weight (counted as 0 kg: ${unknown.map((s) => s.label).slice(0, 3).join(', ')}${unknown.length > 3 ? ', ...' : ''}), so the payload of ${Math.round(l.capacity.kg)} kg is not fully checked.`,
          });
        }
      }
      for (const s of l.stops) {
        const start = s.serviceStartMin;
        const hs = s.hardStartMin;
        const he = s.hardEndMin;
        const outside =
          s.hardWindowOk === false ||
          (start !== null && ((hs !== undefined && hs !== null && start < hs) || (he !== undefined && he !== null && start > he)));
        if (outside) {
          const w = hs !== undefined || he !== undefined ? ` ${hs === null || hs === undefined ? '--:--' : hhmm(hs)}-${he === null || he === undefined ? '--:--' : hhmm(he)}` : '';
          v({ ...at(l), code: 'HARD_WINDOW', message: `${code} load ${l.loadNo}: ${s.label} is served at ${start === null ? '--:--' : hhmm(start)}, outside its receiving hours${w}.` });
        }
        if (start !== null && s.departureMin !== null && s.departureMin < start) {
          v({ ...at(l), code: 'STOP_TIMES', message: `${code} load ${l.loadNo}: ${s.label} leaves (${hhmm(s.departureMin)}) before it is served (${hhmm(start)}).` });
        }
        if (start !== null && s.etaMin !== null && start < s.etaMin - TOL_MIN) {
          v({ ...at(l), code: 'STOP_TIMES', message: `${code} load ${l.loadNo}: ${s.label} is served (${hhmm(start)}) before the truck arrives (${hhmm(s.etaMin)}).` });
        }
      }
      if (l.returnMin < l.departMin) {
        v({ ...at(l), code: 'STOP_TIMES', message: `${code} load ${l.loadNo} is back (${hhmm(l.returnMin)}) before it leaves (${hhmm(l.departMin)}).` });
      }
    }
    // The truck day, with the rules of the loads (the later load's rules for its turnaround).
    for (let i = 1; i < loads.length; i++) {
      const a = loads[i - 1];
      const b = loads[i];
      if (!b.rules) {
        if (b.departMin < a.returnMin - TOL_MIN) {
          v({ ...at(b), code: 'TURNAROUND', message: `${code} load ${b.loadNo} leaves at ${hhmm(b.departMin)}, before load ${a.loadNo} is back (${hhmm(a.returnMin)}).`, shortBy: a.returnMin - b.departMin });
        }
        continue;
      }
      const need = b.rules.reloadMin + b.rules.loadingMinPerCase * b.cases;
      const ready = a.returnMin + need;
      if (b.departMin < ready - TOL_MIN) {
        v({
          ...at(b),
          code: 'TURNAROUND',
          message: `${code} load ${b.loadNo} leaves at ${hhmm(b.departMin)}, but after load ${a.loadNo} (back ${hhmm(a.returnMin)}) the truck needs ${r1(need)} min to reload and load ${b.cases} cases: ready ${hhmm(ready)}.`,
          shortBy: r1(ready - b.departMin),
        });
      }
    }
    const rules = [...loads].reverse().find((l) => l.rules)?.rules ?? null;
    if (rules) {
      const first = loads[0];
      const earliest = Math.max(rules.shiftStartMin, rules.depotOpenMin, rules.availableFromMin ?? 0);
      if (first.departMin < earliest - TOL_MIN) {
        const why = earliest === rules.shiftStartMin ? 'the shift start' : earliest === rules.depotOpenMin ? 'the depot opening' : "the truck's availability";
        v({ ...at(first), code: 'EARLY_DEPARTURE', message: `${code} load ${first.loadNo} leaves at ${hhmm(first.departMin)}, before ${why} (${hhmm(earliest)}).`, shortBy: earliest - first.departMin });
      }
      for (const l of loads) {
        if (l.returnMin > rules.depotCloseMin + TOL_MIN) {
          v({ ...at(l), code: 'DEPOT_CLOSE', message: `${code} load ${l.loadNo} is back at ${hhmm(l.returnMin)}, after the depot closes (${hhmm(rules.depotCloseMin)}).`, shortBy: l.returnMin - rules.depotCloseMin });
        }
        if (rules.availableToMin !== null && l.returnMin > rules.availableToMin + TOL_MIN) {
          v({ ...at(l), code: 'TRUCK_AVAILABILITY', message: `${code} load ${l.loadNo} is back at ${hhmm(l.returnMin)}, after the truck's availability ends (${hhmm(rules.availableToMin)}).`, shortBy: l.returnMin - rules.availableToMin });
        }
      }
      const startDay = Math.min(...loads.map((l) => l.departMin));
      const endDay = Math.max(...loads.map((l) => l.returnMin));
      if (endDay - startDay > rules.shiftMaxMin + TOL_MIN) {
        const last = loads[loads.length - 1];
        v({ ...at(last), code: 'SHIFT_LIMIT', message: `${code} is out from ${hhmm(startDay)} to ${hhmm(endDay)}: longer than the ${Math.floor(rules.shiftMaxMin / 60)}h${String(rules.shiftMaxMin % 60).padStart(2, '0')} shift maximum.`, shortBy: endDay - startDay - rules.shiftMaxMin });
      }
      if (loads.length > rules.maxTrips) {
        v({ ...at(loads[loads.length - 1]), code: 'TRIPS', message: `${code} has ${loads.length} loads; at most ${rules.maxTrips} per day.` });
      }
    }
  }

  // The solver's own report for the option in use.
  const solverStatus = input.solver ? input.solver.status : 'UNKNOWN';
  for (const sv of input.solver?.violations ?? []) {
    const truckId = sv.truck_id ?? null;
    const load = truckId !== null ? input.loads.find((l) => l.truckId === truckId && l.loadNo === sv.load_no) : undefined;
    const past = !!load && load.onRoad;
    out.push({
      code: sv.code,
      severity: past ? 'WARN' : 'BLOCK',
      source: 'SOLVER',
      truckId,
      truckCode: truckId !== null ? (codeOf.get(truckId) ?? truckId) : null,
      loadId: load?.id ?? null,
      loadNo: sv.load_no ?? null,
      message: past ? sv.message + history : sv.message,
      shortBy: sv.short_by_min ?? null,
    });
  }

  const trucks: PlanFeasibility['trucks'] = {};
  const globalBlocks = out.filter((x) => x.severity === 'BLOCK' && x.truckId === null).length;
  for (const [truckId, list] of byTruck) {
    const mine = out.filter((x) => x.truckId === truckId);
    const blocking = mine.filter((x) => x.severity === 'BLOCK').length + globalBlocks;
    const status: TruckTiming = blocking
      ? 'VIOLATED'
      : solverStatus === 'VERIFIED' || solverStatus === 'VIOLATED'
        ? 'VERIFIED' // the solver's violations are per truck: none on this one
        : solverStatus === 'UNVERIFIED'
          ? 'UNVERIFIED'
          : 'STRUCTURAL_ONLY';
    trucks[truckId] = { truckCode: list[0].truckCode, status, ok: status === 'VERIFIED' || status === 'STRUCTURAL_ONLY', blocking, warnings: mine.length - (blocking - globalBlocks) };
  }
  const all = Object.values(trucks);
  const status: TruckTiming = all.some((t) => t.status === 'VIOLATED')
    ? 'VIOLATED'
    : all.some((t) => t.status === 'UNVERIFIED')
      ? 'UNVERIFIED'
      : solverStatus === 'UNKNOWN'
        ? 'STRUCTURAL_ONLY'
        : 'VERIFIED';
  return {
    v: FEASIBILITY_VERSION,
    ok: all.every((t) => t.ok) && globalBlocks === 0,
    status,
    source: input.solver ? 'SOLVER_AND_WEB' : 'LEGACY_STRUCTURAL_ONLY',
    solverStatus,
    solverTiming: input.solver?.timing ?? null,
    trucks,
    violations: out,
    inputHash: inputHash(input),
    checkedAt: now.toISOString(),
  };
}

/** A stored RunPlan.feasibilityJson, or null. */
export function readFeasibility(json: unknown): PlanFeasibility | null {
  const f = json as Partial<PlanFeasibility> | null;
  return f && typeof f === 'object' && typeof f.ok === 'boolean' && typeof f.inputHash === 'string' && f.trucks ? (f as PlanFeasibility) : null;
}

/** Operator switch: FEASIBILITY_GATE=warn lets unverified truck-days be locked and dispatched
 * (the violations are still shown and audited). Anything else, or unset, enforces the gate. */
export function feasibilityGateMode(env: Record<string, string | undefined> = process.env): 'enforce' | 'warn' {
  return (env.FEASIBILITY_GATE ?? '').trim().toLowerCase() === 'warn' ? 'warn' : 'enforce';
}

export { truckDayOk, truckViolations, TIMING_TEXT } from './feasibility-view';
